solutions/fully-configurable/variables.tf

```diff
@@ -0,0 +1,501 @@
+
+
+##############################################################################
+# Virtual server instance Variables
+##############################################################################
+
+variable "vsi_name" {
+  description = "The name of the Virtual server instance."
+  type        = string
+  default     = "vsi"
+}
+
+variable "image_id" {
+  description = "Image ID used for Virtual server instance. Run 'ibmcloud is images' to find available images in a region."
+  type        = string
+  nullable    = false
+}
+
+variable "ssh_public_keys" {
+  description = "List of public SSH keys for Virtual server instance creation which does not already exist in the deployment region. Must be an RSA key with a key size of either 2048 bits or 4096 bits (recommended) - See https://cloud.ibm.com/docs/vpc?topic=vpc-ssh-keys. To use an existing keys, select values for the variable `existing_ssh_key_ids` instead. You can also choose to auto generate an ssh key for you by setting `auto_generate_ssh_key` to true or select existing ssh keys created in the cloud using this variable `existing_ssh_key_ids`."
+  type        = list(string)
+  default     = []
+
+  validation {
+    error_message = "Public SSH Key must be a valid ssh rsa public key."
+    condition     = alltrue([for ssh in var.ssh_public_keys : can(regex("ssh-rsa AAAA[0-9A-Za-z+/]+[=]{0,3} ?([^@]+@[^@]+)?", ssh))])
+  }
+
+  validation {
+    condition     = var.auto_generate_ssh_key || length(var.ssh_public_keys) > 0 || length(var.existing_ssh_key_ids) > 0 ? true : false
+    error_message = "Please provide a value for either `ssh_public_keys` or `existing_ssh_key_ids`, or `auto_generate_ssh_key` must be set to true."
+  }
+}
+
+variable "existing_ssh_key_ids" {
+  description = "The IDs of existing SSH keys to use while creating Virtual server instance. You can also choose to auto generate an ssh key for you by setting `auto_generate_ssh_key` to true or provide a list of ssh public keys in `ssh_public_keys` for private ssh keys own."
+  type        = list(string)
+  default     = []
+}
+
+variable "auto_generate_ssh_key" {
+  description = "An SSH key pair (a public and private key) is automatically generated for you. The private key is outputted as an sensitive value which can be stored in the secret mananger. The public key is stored in your VPC and you can download it from the SSH key details page. Alternately, if you want to bring your own ssh keys you either select the existing ssh keys created in the cloud using this variable `existing_ssh_key_ids` or provide a list of ssh public keys in `ssh_public_keys` for private ssh keys own."
+  type        = bool
+  default     = true
+  nullable    = false
+}
+
+variable "machine_type" {
+  description = "The Virtual server instance machine type. Run 'ibmcloud is instance-profiles' to get a list of regional profiles."
+  type        = string
+  default     = "cx2-2x4"
```

Comment:
If the profiles are regional, then this default value will fail in some regions right? Perhaps this should be a required input so the user experience is better? Otherwise the consumer will only fail after terraform plan or apply has occurred and realise they need to match the profile with the region.